core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/SamplesReceivedWorkflow.java

```diff
@@ -0,0 +1,716 @@
+import java.util.stream.Stream;
+
+import com.google.common.collect.Sets;
+
+import uk.ac.bbsrc.tgac.miso.core.data.Box;
+import uk.ac.bbsrc.tgac.miso.core.data.QcTarget;
+import uk.ac.bbsrc.tgac.miso.core.data.Sample;
+import uk.ac.bbsrc.tgac.miso.core.data.SampleAliquot;
+import uk.ac.bbsrc.tgac.miso.core.data.SampleQC;
+import uk.ac.bbsrc.tgac.miso.core.data.impl.view.BoxableView;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.AbstractWorkflow;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep.InputType;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.WorkflowExecutor;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.WorkflowStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.WorkflowStepPrompt;
+import uk.ac.bbsrc.tgac.miso.core.util.BoxUtils;
+
+public class SamplesReceivedWorkflow extends AbstractWorkflow {
+
+  private final StockStep stockStep = new StockStep();
+  private final QCStep qcStep = new QCStep();
+  private final StockBoxStep stockBoxStep = new StockBoxStep();
+  private StockBoxPositionStep stockBoxPositionStep;
+  private final AliquotStep aliquotStep = new AliquotStep();
+  private final AliquotBoxStep aliquotBoxStep = new AliquotBoxStep();
+  private List<WorkflowStep> aliquotHandlingSteps = Collections.emptyList();
+
+  private final Set<String> occupiedLocations = new HashSet<>();
+
+  @Override
+  protected List<WorkflowStep> getCompletedSteps() {
+    return Stream
+        .concat(Stream.of(stockStep, qcStep, stockBoxStep, stockBoxPositionStep, aliquotStep, aliquotBoxStep),
+            aliquotHandlingSteps.stream())
+        .filter(this::hasInput)
+        .collect(Collectors.toList());
+  }
+
+  private boolean hasInput(WorkflowStep step) {
+    return step != null && step.getProgressStep() != null;
+  }
+
+  @Override
+  protected WorkflowName getWorkflowName() {
+    return WorkflowName.SAMPLES_RECEIVED;
+  }
+
+  @Override
+  public WorkflowStepPrompt getStep(int stepNumber) {
+    if (stepNumber == 0) {
+      return stockStep.getPrompt();
+    } else if (stepNumber == 1) {
+      return qcStep.getPrompt();
+    } else if (stepNumber == 2) {
+      return stockBoxStep.getPrompt();
+    } else if (stepNumber == 3) {
+      return stockBoxPositionStep.getPrompt();
+    } else if (stepNumber == 4) {
+      return aliquotStep.getPrompt();
+    } else if (stepNumber == 5) {
+      return aliquotBoxStep.getPrompt();
+    } else {
+      return aliquotHandlingSteps.get(asAliquotIndex(stepNumber)).getPrompt();
+    }
+  }
+
+  private int asAliquotIndex(int stepNumber) {
+    return stepNumber - 6;
+  }
+
+  @Override
+  public boolean isComplete() {
+    return Stream.concat(Stream.of(stockStep, qcStep, stockBoxStep, stockBoxPositionStep, aliquotStep, aliquotBoxStep),
+        aliquotHandlingSteps.stream()).allMatch(this::hasInput);
+  }
+
+  @Override
+  public List<String> processInput(int stepNumber, ProgressStep step) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)